src/DynamicProjectOptions.cmake

```diff
@@ -0,0 +1,152 @@
+#  * ENABLE_SANITIZER_UNDEFINED: ON for Compilers that support it
+#  * ENABLE_CPPCHECK: ON for Ninja/Makefiles
+
+# For non-developer builds
+#  -DENABLE_DEVELOPER_MODE:BOOL=OFF
+# Is recommended
+
+# In developer mode, all features have options that show up in the CMake GUI tools
+
+# dynamic_project_options() macro enables all recommended defaults with appropriately
+# applied options from the GUI which are set
+
+# Any default can be overridden
+# set(<feature_name>_DEFAULT <value>) - set default for both user and developer modes
+# set(<feature_name>_DEVELOPER_DEFAULT <value>) - set default for developer mode
+# set(<feature_name>_USER_DEFAULT <value>) - set default for user mode
+
+option(ENABLE_DEVELOPER_MODE "Set up defaults for a developer of the project, and let developer change options" ON)
+
+if((CMAKE_CXX_COMPILER_ID MATCHES ".*Clang.*" OR CMAKE_CXX_COMPILER_ID MATCHES ".*GNU.*") AND NOT WIN32)
+  set(SUPPORTS_UBSAN ON)
+else()
+  set(SUPPORTS_UBSAN OFF)
+endif()
+
+if((CMAKE_CXX_COMPILER_ID MATCHES ".*Clang.*" OR CMAKE_CXX_COMPILER_ID MATCHES ".*GNU.*") AND WIN32)
+  set(SUPPORTS_ASAN OFF)
+else()
+  set(SUPPORTS_ASAN ON)
+endif()
+
+# ccache, clang-tidy, cppcheck are only supported with Ninja and Makefile based generators
+# note that it is possible to use Ninja with cl, so this still allows clang-tidy on Windows
+# with CL.
+#
+# We are only setting the default options here. If the user attempts to enable
+# these tools on a platform with unknown support, they are on their own.
+#
+# Also note, cppcheck has an option to be run on VCproj files, so we should investigate that
+# Further note: MSVC2022 has builtin support for clang-tidy, but I can find
+# no way to enable that via CMake
+if(CMAKE_GENERATOR MATCHES ".*Makefile*." OR CMAKE_GENERATOR MATCHES ".*Ninja*")
+  set(MAKEFILE_OR_NINJA ON)
+else()
+  set(MAKEFILE_OR_NINJA OFF)
+endif()
+
+include(CMakeDependentOption)
+
+# <option name>;<user mode default>;<developer mode default>;<description>
+set(options
+    "ENABLE_CACHE\;${MAKEFILE_OR_NINJA}\;${MAKEFILE_OR_NINJA}\;Enable ccache on Unix"
+    "WARNINGS_AS_ERRORS\;OFF\;ON\;Treat warnings as Errors"
+    "ENABLE_CLANG_TIDY\;OFF\;${MAKEFILE_OR_NINJA}\;Enable clang-tidy analysis during compilation"
+    "ENABLE_CONAN\;ON\;ON\;Automatically integrate Conan for package management"
+    "ENABLE_COVERAGE\;OFF\;OFF\;Analyze and report on coverage"
+    "ENABLE_SANITIZER_ADDRESS\;OFF\;${SUPPORTS_ASAN}\;Make memory errors into hard runtime errors (windows/linux/macos)"
+    "ENABLE_SANITIZER_UNDEFINED_BEHAVIOR\;OFF\;${SUPPORTS_UBSAN}\;Make certain types (numeric mostly) of undefined behavior into runtime errors"
+    "ENABLE_CPPCHECK\;OFF\;${MAKEFILE_OR_NINJA}\;Enable cppcheck analysis during compilation"
+    "ENABLE_IPO\;OFF\;OFF\;Enable whole-program optimization"
+    "ENABLE_INCLUDE_WHAT_YOU_USE\;OFF\;OFF\;Enable include-what-you-use analysis during compilation"
+    "ENABLE_PCH\;OFF\;OFF\;Enable pre-compiled-headers support"
```

Comment:
Should we automate this and exclude mingw on Windows?

Reply:
I don't want to 100% exclude any compiler from any setting, in case things change in the near future, if that is what you mean. Just more like "if you enable something that's not the default, then you're on your own" was my philosophy here.

Reply:
I meant setting the default like the Address sanitizer.

Reply:
Sorry, I'm not sure what you're asking. I did automate the defaults for things like ASAN on Windows, etc, up above this section:

https://github.com/cpp-best-practices/cpp_starter_project/pull/181/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR95-R100

Reply:
I am saying that we can automate PCH's default like how we did for ASAN

```
    "ENABLE_PCH\;OFF\;${NOT_WIN32_AND_GNU}\;Enable pre-compiled-headers support"
```

Reply:
Oh, OK, but I would reverse that. On for User, off for Developer. Just to speed up build times for the user and reduce the risk of accidentally including headers during development.

I went Off/Off ultimately since you had expressed some discomfort with it, and it was low priority to me.